calour/training.py

```diff
@@ -121,24 +126,202 @@ def split_train_test(exp: Experiment, field, test_size, train_size=None, stratif
-def collect_cv_prediction(exp: Experiment, field, estimator, cv, scoring, inplace=False):
-    '''Do the CV
+def classify(exp: Experiment, field, estimator, cv=RepeatedStratifiedKFold(3, 1),
+             predict='predict_proba', param_grid=None):
+    '''Evaluate classification during cross validation.
+
+    Parameters
+    ----------
+    field : str
+        column name in the sample metadata, which contains the classes we want to predict.
+    estimator : object
+        scikit-learn estimator. e.g. :class:`sklearn.ensemble.RandomForestClassifer`
+    cv : int, cross-validation generator or an iterable
+        similar to the `cv` parameter in :class:`sklearn.model_selection.GridSearchCV`
+    predict : {'predict', 'predict_proba'}
+        the method to predict the validation sets
+    param_grid : dict of string to sequence, or sequence of such
+        the parameter passing to :class:`sklearn.model_selection.ParameterGrid`
 
+    Yields
+    ------
+    pandas.DataFrame
+        The result of prediction per sample for a given parameter set. It contains the
+        following columns:
+        - Y_TRUE: the true class for the samples
+        - SAMPLE: sample IDs
+        - CV: which split of the cross validation
+        - Y_PRED: the predicted class for the samples (if "predict")
+        - mutliple columns with each contain probabilities predicted as each class (if "predict_proba")
     '''
-    # from sklearn.model_selection._split import check_cv
-    # from sklearn.base import is_classifier
-    # logger.debug('')
-    # if inplace:
-    #     new = exp
-    # else:
-    #     new = exp.copy()
-    # cv = check_cv(cv, y, classifier=is_classifier(estimator))
-    # for params in paramgrid:
-    #     for train_x, train_y in cv:
-    #         estimator.fit(train_x, train_y)
-
-    # return yobs, yhat, model
+    X = exp.data
+    y = exp.sample_metadata[field]
+    cv = check_cv(cv, y, classifier=is_classifier(estimator))
+
+    if param_grid is None:
+        # use sklearn default param values for the given estimator
+        param_grid = {}
+
+    for param in ParameterGrid(param_grid):
+        logger.debug('run classification with parameters: %r' % param)
+        dfs = []
+        for i, (train, test) in enumerate(cv.split(X, y)):
+            model = estimator(**param)
+            model.fit(X[train], y[train])
+            pred = getattr(model, predict)(X[test])
+            if pred.ndim > 1:
+                df = pd.DataFrame(pred, columns=model.classes_)
+            else:
+                df = pd.DataFrame(pred, columns=['Y_PRED'])
+            df['Y_TRUE'] = y[test].values
+            df['SAMPLE'] = y[test].index.values
+            df['CV'] = i
+            dfs.append(df)
+        yield pd.concat(dfs, axis=0).reset_index(drop=True)
+
+
+def plot_cm(result, normalize=False, title='confusion matrix', cmap=plt.cm.Blues, ax=None):
+    '''Plot confusion matrix
+
+    Parameters
+    ----------
+    result : pandas.DataFrame
```

Comment:
maybe change parameter name to classify_df or something else?
result as an input is a bit confusing...

Reply:
i don't think classify_df is better. updated docstring, which should make it clearer